src/Voron/Platform/Win32/WindowsMemoryMapPager.cs

```diff
@@ -410,26 +475,124 @@ public override void ReleaseAllocationInfo(byte* baseAddress, long size)
             NativeMemory.UnregisterFileMapping(_fileInfo.FullName, new IntPtr(baseAddress), size);
         }
 
-        public override void MaybePrefetchMemory(List<long> pagesToPrefetch)
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int GetSegmentState(long segment)
+        {
+            if (segment < 0)
+                return AlreadyPrefetch;
+
+            byte value = this._prefetchTable[segment / 2];
+            if (segment % 2 == 0)
+            {
+                // The actual value is in the high byte.
+                value = (byte)(value >> EvenPrefetchMaskShift);
+            }
+            else
+            {
+                value = (byte)(value & OddPrefetchCountMask);
+            }
+
+            return value;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetSegmentState(long segment, int state)
+        {
+            byte value = this._prefetchTable[segment / 2];
+            if (segment % 2 == 0)
+            {
+                // The actual value is in the high byte.
+                value = (byte)((value & OddPrefetchCountMask) | (state << EvenPrefetchMaskShift));
+            }
+            else
+            {
+                value = (byte)((value & EvenPrefetchCountMask) | state);
+            }
+
+            this._prefetchTable[segment / 2] = value;
+        }
+
+        private void ResetPrefetchTable()
+        {
+            this._refreshCounter = 0;
+
+            // We will zero out the whole table to reset the prefetching behavior. 
+            for (int i = 0; i < this._prefetchTable.Length; i++)
+                this._prefetchTable[i] = 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void MaybePrefetchSegment(byte* baseAddress, long segment)
+        {
+            Win32MemoryMapNativeMethods.WIN32_MEMORY_RANGE_ENTRY entry;
+            entry.NumberOfBytes = (IntPtr)PrefetchSegmentSize;
+            entry.VirtualAddress = baseAddress + segment * PrefetchSegmentSize;
+
+            Win32MemoryMapNativeMethods.PrefetchVirtualMemory(Win32Helper.CurrentProcess, (UIntPtr)1, &entry, 0);
```

Comment:
Why do it one item at a time?

Reply:
Because that is only used for the case of acquiring a pointer, so there is no 'context' beyond someone else has asked for this one before too.